### PR TITLE
#1270 doc preview only for PR for master [WIP]

### LINKED ISSUE #1270: Preview of built docs in the PR
## 🚀 Feature

Idea is to be able to preview built docs on the CI to simplify PR review.
This issue may be a bit tricky to solve, here some guidelines:
- we build our docs with travis ci : https://github.com/pytorch/ignite/blob/master/.travis.yml#L62
- find out how we could keep built docs as artifacts and make them accessible after the build
- update .travis.yml for that
- update https://github.com/pytorch/ignite/blob/master/.github/PULL_REQUEST_TEMPLATE.md with some details on how to fetch the link of built docs.

For Hacktoberfest contributors, feel free to ask questions for details if any and say that you would like to tackle the issue.
Please, take a look at [CONTRIBUTING guide](https://github.com/pytorch/ignite/blob/master/CONTRIBUTING.md).

### PR DESCRIPTION
**Pull request by sroy8091**
Fixes #1270

Description:
update of travis for doc preview

Check list:
* [ ] Documentation is updated (if required)

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -62,7 +62,7 @@ jobs:
     # GitHub Pages Deployment: https://docs.travis-ci.com/user/deployment/pages/
     - stage: Docs
       python: "3.7"
-      if: type != cron
+      if: type != cron AND branch = master
```
Comment (vfdev-5) [inline .travis.yml:65]: I'm not sure if this worked. See current travis CI report where we still build the docs.

**Review by vfdev-5 — COMMENTED** [review]
Thanks for the PR !
Could you also add ["Nice to have"](https://github.com/pytorch/ignite/issues/1270#issuecomment-703180980) features ?

**Comment by sroy8091** [comment]
@vfdev-5 as you can see the travis build it doesn't have doc build so only if pr base branch is master it will create. Is there a different behaviour you are expecting?

**Comment by vfdev-5** [comment]
Thanks @sroy8091 ! Probably, I was not enough clear about the intention. We would like to have the following behaviour:
- if build is a PR, we do not build the docs as netlify will do it and show a preview.
- if it is a merged PR and we rebuild the master, then we build and deploy the docs as previously.

**Comment by vfdev-5** [comment]
@sroy8091 any updates on this PR ? Let me know if you get stuck at some point such that we could help.

**Comment by sroy8091** [comment]
@vfdev-5 yes i tried different things but it seems to not working. Does this seems workable https://stackoverflow.com/questions/12878568/travis-ci-triggers-build-only-for-master-branch-in-github-repo?

**Comment by vfdev-5** [comment]
@sroy8091 can you check travis docs : https://docs.travis-ci.com/user/pull-requests#pull-requests-and-security-restrictions
and see if we can use `TRAVIS_PULL_REQUEST` as a condition.

That's correct that `branch: onlyl` specifies PR targeting a branch

> To only build pull requests targeting specific branches you can use the branches: only: key, which will also restrict the branches that trigger builds.

**Comment by sroy8091** [comment]
@vfdev-5 now we need to set this variable so that doc will build in actual envs

**Comment by vfdev-5** [comment]
@sroy8091 thanks for the update. Looks good.


Following https://github.com/pytorch/ignite/pull/1350#pullrequestreview-501617852
> Could you also add "Nice to have" features ?

How about this ?

**Comment by sroy8091** [comment]
@vfdev-5 can you check why i am getting pytest error in travis build?

File: .travis.yml [hunk 2]
```diff
@@ -63,18 +50,17 @@ jobs:
...
 
-      install:
-        - pip install -r docs/requirements.txt
-        - pip install git+https://github.com/vfdev-5/sphinxcontrib-versioning.git
-
       script:
-        - sphinx-versioning --use-master-conf --use-master-templates build --greatest-tag --whitelist-branches master docs/source docs/build/html
-        # Create .nojekyll file to serve correctly _static and friends
-        - touch docs/build/html/.nojekyll
+        - chmod +x .travis/install_docs.sh
```
Comment (vfdev-5) [inline .travis.yml:59]: Typo as no `install_docs.sh`, right ?

File: .travis.yml [hunk 3]
```diff
@@ -63,18 +50,17 @@ jobs:
     # GitHub Pages Deployment: https://docs.travis-ci.com/user/deployment/pages/
     - stage: Docs
       python: "3.7"
-      if: type != cron
+
+#      if: type != cron and $TRAVIS_PULL_REQUEST = True
       # Use previously defined before_install
       before_install: *before_install
 
-      install:
```
Comment (vfdev-5) [inline .travis.yml:70]: Let's keep the same step:
```
install:
    - chmod +x .travis/install_docs.sh
    - ./.travis/install_docs_deps.sh
```
and 
```
script:
  - ...
  - ...
```

File: .travis/install_deps.sh [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+wget https://repo.continuum.io/miniconda/Miniconda3-latest-Linux-x86_64.sh -O miniconda.sh;
+bash miniconda.sh -b -p $HOME/miniconda
+export PATH="$HOME/miniconda/bin:$PATH"
+hash -r
+conda config --set always_yes yes --set changeps1 no
+conda update -q conda
+  # Useful for debugging any issues with conda
+conda info -a
+conda create -q -n test-environment pytorch cpuonly python=$TRAVIS_PYTHON_VERSION -c $PYTORCH_CHANNEL
+source activate test-environment
```
Comment (vfdev-5) [inline .travis/install_deps.sh:12]: `source` wont work externally, so when you call pytest etc conda env is not activated.
I propose to change that and install everything in `base` conda env. Thus, we do not have to create/activate another env.

File: .travis.yml [hunk 5]
```diff
@@ -20,25 +20,12 @@ stages:
...
-  # - if [[ "$PYTORCH_CHANNEL" == "pytorch-nightly" ]]; then pip install --upgrade git+https://github.com/pytorch/vision.git; else conda install torchvision cpuonly python=$TRAVIS_PYTHON_VERSION -c $PYTORCH_CHANNEL; fi
-  - conda install torchvision cpuonly python=$TRAVIS_PYTHON_VERSION -c $PYTORCH_CHANNEL
-  # Install all test/examples dependencies
-  - pip install -r requirements-dev.txt
+  - chmod +x .travis/install_deps.sh
+  - ./.travis/install_deps.sh
 
 install:
   - pip install "mypy==0.782"
+  - pip install pytest-cov
```
Comment (vfdev-5) [inline .travis.yml:28]: normally, this is not required.

**Review by vfdev-5 — COMMENTED** [review]
@sroy8091 thanks for the update. I left few comments on that.

File: .travis.yml [hunk 6]
```diff
@@ -20,22 +20,8 @@ stages:
...
-  - conda info -a
-  - conda create -q -n test-environment pytorch cpuonly python=$TRAVIS_PYTHON_VERSION -c $PYTORCH_CHANNEL
-  - source activate test-environment
-  # Keep fix in case of problem with torchvision nightly releases
-  # - if [[ "$PYTORCH_CHANNEL" == "pytorch-nightly" ]]; then pip install --upgrade git+https://github.com/pytorch/vision.git; else conda install torchvision cpuonly python=$TRAVIS_PYTHON_VERSION -c $PYTORCH_CHANNEL; fi
-  - conda install torchvision cpuonly python=$TRAVIS_PYTHON_VERSION -c $PYTORCH_CHANNEL
-  # Install all test/examples dependencies
-  - pip install -r requirements-dev.txt
+  - chmod +x .travis/install_deps.sh
+  - ./.travis/install_deps.sh
```
Comment (vfdev-5) [inline .travis.yml:24]: Can you try using `sh .travis/install_deps.sh` instead of 2 lines?

File: .travis.yml [hunk 7]
```diff
@@ -63,18 +49,19 @@ jobs:
...
+
+#      if: type != cron and $TRAVIS_PULL_REQUEST = True
       # Use previously defined before_install
       before_install: *before_install
 
       install:
-        - pip install -r docs/requirements.txt
-        - pip install git+https://github.com/vfdev-5/sphinxcontrib-versioning.git
+        - chmod +x .travis/install_docs_deps.sh
+        - ./.travis/install_docs_deps.sh
```
Comment (vfdev-5) [inline .travis.yml:59]: Same here and below

File: .travis/install_deps.sh [hunk 8]
```diff
@@ -0,0 +1,15 @@
...
+bash miniconda.sh -b -p $HOME/miniconda
+export PATH="$HOME/miniconda/bin:$PATH"
+hash -r
+conda config --set always_yes yes --set changeps1 no
+conda update -q conda
+  # Useful for debugging any issues with conda
+conda info -a
+# Keep fix in case of problem with torchvision nightly releases
+# - if [[ "$PYTORCH_CHANNEL" == "pytorch-nightly" ]]; then pip install --upgrade git+https://github.com/pytorch/vision.git; else conda install torchvision cpuonly python=$TRAVIS_PYTHON_VERSION -c $PYTORCH_CHANNEL; fi
+conda install torchvision
```
Comment (vfdev-5) [inline .travis/install_deps.sh:13]: This line is incorrect.

Let's remove all the lines 11 - 13 and instead put :
```
conda install torchvision cpuonly python=$TRAVIS_PYTHON_VERSION -c $PYTORCH_CHANNEL
```

File: .travis.yml [hunk 9]
```diff
@@ -63,18 +49,19 @@ jobs:
     # GitHub Pages Deployment: https://docs.travis-ci.com/user/deployment/pages/
     - stage: Docs
       python: "3.7"
-      if: type != cron
+
+#      if: type != cron and $TRAVIS_PULL_REQUEST = True
```
Comment (vfdev-5) [inline .travis.yml:53]: This should be uncommented, right ? And the condition probably wont work neither, please see the docs : https://docs.travis-ci.com/user/conditions-v1#integration

**Review by vfdev-5 — CHANGES_REQUESTED** [review]
@sroy8091 thanks for updating the PR, but it still does not look correct.

**Comment by sroy8091** [comment]
Yeah, still working on it. The error in build is related to pytest. Is there any way to test this locally?

**Comment by vfdev-5** [comment]
>  The error in build is related to pytest. Is there any way to test this locally?

@sroy8091 if you have a docker you can try to run step by step CI intructions in ubuntu:18.04 container. This should be rather close to what it is in travis, except all env vars like `TRAVIS_*`.

**Comment by vfdev-5** [comment]
Closed in favor of #1457

**Pull request closed without merging**